### ISSUE TITLE
Buffer mismatch occured in bdevperf with verify

### ISSUE THREAD

**Issue by shuhei-matsumoto** [spdk/spdk#1208]
The following was observed in CI.

```
************************************
START TEST bdev_verify
************************************
 18:08:42	# xtrace_enable
 18:08:42	# /var/jenkins/workspace/ubuntu18-autotest/spdk/test/bdev/bdevperf/bdevperf -c /var/jenkins/workspace/ubuntu18-autotest/spdk/test/bdev/bdev.conf -q 128 -o 4096 -w verify -t 5
Starting SPDK v20.04-pre git sha1 9cce892b4f / DPDK 19.11.0 initialization...
[ DPDK EAL parameters: bdevperf --no-shconf -c 0x1 --log-level=lib.eal:6 --log-level=lib.cryptodev:5 --log-level=user1:6 --iova-mode=pa --base-virtaddr=0x200000000000 --match-allocations --file-prefix=spdk_pid30913 ]
EAL: No available hugepages reported in hugepages-1048576kB
app.c: 642:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 316:_spdk_reactor_run: *NOTICE*: Reactor started on core 0
cryptodev_aesni_mb_create() line 1321: IPSec Multi-buffer library version used: 0.52.0

vbdev_passthru.c: 533:vbdev_passthru_init: *NOTICE*: conf parse matched: Malloc3
vbdev_passthru.c: 629:vbdev_passthru_register: *NOTICE*: Match on Malloc3
vbdev_passthru.c: 672:vbdev_passthru_register: *NOTICE*: io_device created at: 0x0x615000002880
vbdev_passthru.c: 684:vbdev_passthru_register: *NOTICE*: bdev opened
vbdev_passthru.c: 696:vbdev_passthru_register: *NOTICE*: bdev claimed
vbdev_passthru.c: 709:vbdev_passthru_register: *NOTICE*: pt_bdev registered
vbdev_passthru.c: 710:vbdev_passthru_register: *NOTICE*: created pt_bdev for: TestPT
Running I/O for 5 seconds...
app.c: 712:spdk_app_stop: *WARNING*: spdk_app_stop'd on non-zero
Buffer mismatch! Disk Offset: 4984
Buffer mismatch! Disk Offset: 4986
Buffer mismatch! Disk Offset: 4988
Buffer mismatch! Disk Offset: 4990
Buffer mismatch! Disk Offset: 4992
Buffer mismatch! Disk Offset: 4994
Buffer mismatch! Disk Offset: 4996
Buffer mismatch! Disk Offset: 4998
  18:08:49	# trap - ERR
  18:08:49	# print_backtrace
  18:08:49	# [[ ehxBET =~ e ]]
  18:08:49	# args=("${BASH_ARGV[@]}")
  18:08:49	# local args
  18:08:49	# xtrace_disable
  18:08:49	# '[' no '!=' yes ']'
  18:08:49	# PREV_BASH_OPTS=ehxBET
  18:08:49	# [[ ehxBET == *\x* ]]
  18:08:49	# XTRACE_DISABLED=yes
  18:08:49	# set +x
========== Backtrace start: ==========

```

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://dqtibwqq6s6ux.cloudfront.net/public_build/autotest-per-patch_3313.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_4132.html

**Comment by peluse** [comment]
I can repro this locally with a simplified config.  In debug...

**Comment by peluse** [comment]
I don't believe what I found is the root cause here or it'd fail every time, but I did post one verify fix here https://review.spdk.io/gerrit/c/spdk/spdk/+/1069 
Will continue debug as I was able to repro this one time w/o modifying the config but immediately went to attempting to simplify the scenario after I did.

**Comment by peluse** [comment]
FYI added some prints to see which target the mismatch happened on and what the first dword mismatch is https://review.spdk.io/gerrit/c/spdk/spdk/+/1070/1

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_5066.html

**Comment by shuhei-matsumoto** [comment]
Hi @peluse 

Thanks for the great analysis and fix. For the complete fix, if we use verify, do you think we have to
*  set queue depth to 1, or 
*  use spdk_bdev_comparev_and_writev_blocks?

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_5436.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_5530.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_5534.html

**Comment by shuhei-matsumoto** [comment]
Copied from Slack.

 I saw this issue only twice, and these had the same signature. The log showed mismatch occurred at LBA 4984, 4986, 4988, 4990, and IO size of bdevperf was 4096. So it is very likely that the issue was caused by using AIO bdev. AIO file was block size = 2048 and number of blocks = 5000. So The LBAs was almost end of the LBA file.
Queue depth was 128 and so as you think, your fix may not be for the current issue

**Comment by jimharris** [comment]
> FYI added some prints to see which target the mismatch happened on and what the first dword mismatch is https://review.spdk.io/gerrit/c/spdk/spdk/+/1070/1

This patch should help get closer to the bottom of this.  We'll get this reviewed and merged.

@shuhei-matsumoto also suggested trying a patch with multiple AIO bdevs to see if that accelerates the problem.  We might not want to check it in, but could run it multiple times through the test pool to see if it fails.

**Comment by peluse** [comment]
@jimharris cool, I'll try it

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_5720.html

**Comment by shuhei-matsumoto** [comment]
> Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_5720.html

@peluse @jimharris My anylysis for log was totally wrong. Sorry for that.
New log showed the mismatch occurred offset 0.
I'm suspecting my recent patches caused the issue and investigating it.

 ```
03:17:38	# /var/jenkins/workspace/nvme-autotest/spdk/test/bdev/bdevperf/bdevperf.py perform_tests
Running I/O for 4 seconds...
Buffer mismatch! Disk Offset: 0
Buffer mismatch! Disk Offset: 1
Buffer mismatch! Disk Offset: 2
Buffer mismatch! Disk Offset: 3
Buffer mismatch! Disk Offset: 4
Buffer mismatch! Disk Offset: 5
Buffer mismatch! Disk Offset: 6
Buffer mismatch! Disk Offset: 7
Buffer mismatch! Disk Offset: 8
Buffer mismatch! Disk Offset: 9
Buffer mismatch! Disk Offset: 10
Buffer mismatch! Disk Offset: 11
Buffer mismatch! Disk Offset: 12
Buffer mismatch! Disk Offset: 13
Buffer mismatch! Disk Offset: 14
Buffer mismatch! Disk Offset: 15
Buffer mismatch! Disk Offset: 16
Buffer mismatch! Disk Offset: 17
Buffer mismatch! Disk Offset: 18
Buffer mismatch! Disk Offset: 19
Buffer mismatch! Disk Offset: 20
Buffer mismatch! Disk Offset: 21
Buffer mismatch! Disk Offset: 22
Buffer mismatch! Disk Offset: 23
Buffer mismatch! Disk Offset: 24
Buffer mismatch! Disk Offset: 25
Buffer mismatch! Disk Offset: 26
Buffer mismatch! Disk Offset: 27
Buffer mismatch! Disk Offset: 28
Buffer mismatch! Disk Offset: 29
Buffer mismatch! Disk Offset: 30
Buffer mismatch! Disk Offset: 31
Buffer mismatch! Disk Offset: 32
Buffer mismatch! Disk Offset: 33
Buffer mismatch! Disk Offset: 34
Buffer mismatch! Disk Offset: 35
Buffer mismatch! Disk Offset: 36
Buffer mismatch! Disk Offset: 37
Buffer mismatch! Disk Offset: 38
Buffer mismatch! Disk Offset: 39
Buffer mismatch! Disk Offset: 40
Buffer mismatch! Disk Offset: 41
Buffer mismatch! Disk Offset: 42
Buffer mismatch! Disk Offset: 43
Buffer mismatch! Disk Offset: 44
Buffer mismatch! Disk Offset: 45
Buffer mismatch! Disk Offset: 46
Buffer mismatch! Disk Offset: 47
Buffer mismatch! Disk Offset: 48
Buffer mismatch! Disk Offset: 49
Buffer mismatch! Disk Offset: 50
Buffer mismatch! Disk Offset: 51
Buffer mismatch! Disk Offset: 52
Buffer mismatch! Disk Offset: 53
Buffer mismatch! Disk Offset: 54
Buffer mismatch! Disk Offset: 55
Buffer mismatch! Disk Offset: 56
Buffer mismatch! Disk Offset: 57
Buffer mismatch! Disk Offset: 58
Buffer mismatch! Disk Offset: 59
Buffer mismatch! Disk Offset: 60
Buffer mismatch! Disk Offset: 61
Buffer mismatch! Disk Offset: 62
Buffer mismatch! Disk Offset: 63
Buffer mismatch! Disk Offset: 64
Buffer mismatch! Disk Offset: 65
Buffer mismatch! Disk Offset: 66
Buffer mismatch! Disk Offset: 67
Buffer mismatch! Disk Offset: 68
Buffer mismatch! Disk Offset: 69
Buffer mismatch! Disk Offset: 70
Buffer mismatch! Disk Offset: 71
Buffer mismatch! Disk Offset: 72
Buffer mismatch! Disk Offset: 73
Buffer mismatch! Disk Offset: 74
Buffer mismatch! Disk Offset: 75
Buffer mismatch! Disk Offset: 76
Buffer mismatch! Disk Offset: 77
Buffer mismatch! Disk Offset: 78
Buffer mismatch! Disk Offset: 79
Buffer mismatch! Disk Offset: 80
Buffer mismatch! Disk Offset: 81
Buffer mismatch! Disk Offset: 82
Buffer mismatch! Disk Offset: 83
Buffer mismatch! Disk Offset: 84
Buffer mismatch! Disk Offset: 85
Buffer mismatch! Disk Offset: 86
Buffer mismatch! Disk Offset: 87
Buffer mismatch! Disk Offset: 88
Buffer mismatch! Disk Offset: 89
Buffer mismatch! Disk Offset: 90
Buffer mismatch! Disk Offset: 91
Buffer mismatch! Disk Offset: 92
Buffer mismatch! Disk Offset: 93
Buffer mismatch! Disk Offset: 94
Buffer mismatch! Disk Offset: 95
Buffer mismatch! Disk Offset: 96
Buffer mismatch! Disk Offset: 97
Buffer mismatch! Disk Offset: 98
Buffer mismatch! Disk Offset: 99
Buffer mismatch! Disk Offset: 100
Buffer mismatch! Disk Offset: 101
Buffer mismatch! Disk Offset: 102
Buffer mismatch! Disk Offset: 103
Buffer mismatch! Disk Offset: 104
Buffer mismatch! Disk Offset: 105
Buffer mismatch! Disk Offset: 106
Buffer mismatch! Disk Offset: 107
Buffer mismatch! Disk Offset: 108
Buffer mismatch! Disk Offset: 109
Buffer mismatch! Disk Offset: 110
Buffer mismatch! Disk Offset: 111
Buffer mismatch! Disk Offset: 112
Buffer mismatch! Disk Offset: 113
Buffer mismatch! Disk Offset: 114
Buffer mismatch! Disk Offset: 115
Buffer mismatch! Disk Offset: 116
Buffer mismatch! Disk Offset: 117
Buffer mismatch! Disk Offset: 118
Buffer mismatch! Disk Offset: 119
Buffer mismatch! Disk Offset: 120
Buffer mismatch! Disk Offset: 121
Buffer mismatch! Disk Offset: 122
Buffer mismatch! Disk Ofrequest:
{
  "method": "perform_tests",
  "req_id": 1
}
Got JSON-RPC error response
response:
{
  "code": -32603,
  "message": "bdevperf failed with error Operation not permitted"
}
  03:17:38	# trap - ERR
  03:17:38	# print_backtrace
  03:17:38	# [[ ehxBET =~ e ]]
  03:17:38	# args=("${BASH_ARGV[@]}")
  03:17:38	# local args
  03:17:38	# xtrace_disable
  03:17:38	# '[' no '!=' yes ']'
  03:17:38	# PREV_BASH_OPTS=ehxBET
  03:17:38	# [[ ehxBET == *\x* ]]
  03:17:38	# XTRACE_DISABLED=yes


```

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_5730.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_5736.html

**Comment by shuhei-matsumoto** [comment]
Hi @peluse 

Sorry for confused and wrong information. Buffer mismatch by FTL or OCSSD bdev is not related with this issue. Please ignore FTL or OCSSD cases. I will fix FTL or OCSSD cases during development.

**Comment by peluse** [comment]
proposed fix here: https://review.spdk.io/gerrit/c/spdk/spdk/+/1180

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @tomzawadzki. log: https://ci.spdk.io/public_build/autotest-per-patch_5995.html

**Comment by peluse** [comment]
SOB.  This is also on the same VM that I think all others were reported on.  Will check it out

**Comment by jimharris** [comment]
@tomzawadzki did this last instance happen because the patch wasn’t rebased on top of Paul’s fix?  I noticed you posted this in Gerrit:

The FP would be fixed by rebasing on master from yesterday, but to prevent doing whole series through CI - just posted FP report.

**Comment by peluse** [comment]
@jimharris @tomzawadzki No it had the fix.  I downloaded the patch and confirmed before I re-opened :(

**Comment by peluse** [comment]
thought I had a theory but no... will hold off until @tkulasek confirms rebase/retest - maybe I made a mistake pulling the patch down or something.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @mikeBashStuff. log: https://ci.spdk.io/public_build/autotest-per-patch_6235.html

**Comment by peluse** [comment]
Hmmm, more AIO and nearly all of them on are Ubuntu18. Debug continues...

**Comment by peluse** [comment]
This should do the trick (famous last words) https://review.spdk.io/gerrit/c/spdk/spdk/+/1250

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @shuhei-matsumoto. log: https://ci.spdk.io/public_build/autotest-per-patch_7397.html

**Comment by peluse** [comment]
@shuhei-matsumoto I entered an issue for bdev thread exit active channel thingy here: https://github.com/spdk/spdk/issues/1288 

I ignored the data mismatch until further investigation because it seemed like the mismatch is the result of another error, not a logic error in the w/r/v path.  If you look at the build log for the last one you submitted above, it also has the active channel error.  The other ones I spot checked that drove the fix for the real data miscompare did not.  Also, the one that drove the issue I just pasted in also shows buffer mismatch following the active channel error.

Hope that made sense :) Where are you at with you thinking on 1288? I think that's the bug that should have been cited for the last error and that this issue should be closed (at least for now)
